lib/flavor-downloader.js

```diff
@@ -0,0 +1,140 @@
+/**
+ * Created by AsherS on 8/20/15.
+ */
+
+var Q = require('q');
+var _ = require('underscore');
+var path = require('path');
+var url = require('url');
+var httpUtils = require('./utils/http-utils');
+var qfs = require("q-io/fs");
+var logger = require('./logger/logger');
+var config = require('./Configuration');
+var m3u8Handler = require('./promise-m3u8');
+var chunklistM3u8Generator = require('./ChunklistManifestGenerator');
+
+var FlavorDownloader = (function () {
+
+    function FlavorDownloader(m3u8Url, destPath, entryId, flavor) {
+        this.entryId = entryId;
+        this.flavor = flavor;
+        this.streamUrl = m3u8Url;
+        this.destFolderPath = destPath;
+        this.shouldRun = true;
+        this.pollingInterval = config.get("pollingInterval");
+        this.tmpM3u8Folder = path.join(destPath, 'tmpM3u8');
+        this.playlistName = path.basename(url.parse(m3u8Url).pathname);
+        this.m3u8generator = chunklistM3u8Generator(this.destFolderPath, this.playlistName, config.get("dvrWindow"));
+    }
+
+    var formatLogMsg = function (that, msg) {
+        return "entryId: " + that.entryId + " , flavor: " + that.flavor + " - " + msg;
+    };
+
+    var downloadIndefinitely = function () {
+
+        var that = this;
+        downloadPlaylist(that)
+
+            .fail(function failure(listOfErrors) {
+                logger.error(formatLogMsg(that,"Failed to download all chunks:\n"));
+                var errList = _.map(listOfErrors, function (i) {
+                    return i.reason + "\n";
+                })
+                logger.error(formatLogMsg(that,errList));
+            });
+
+        setTimeout(function () {
+            if (that.shouldRun) {
+                downloadIndefinitely.call(that);
+            }
+            else {
+                logger.info(formatLogMsg(that,'downloader has shut down'));
+            }
+        }, that.pollingInterval);
+    };
+
+    var downloadPlaylist = function (that) {
```

Comment:
you can remove the "that" parameter and call downloadPlaylist with downloadPlaylist.call(this) the same way you do with downloadIndefinitely